### PR TITLE
WIP: BFLATN -> BSATN fast-path for fixed-length rows

### PR DESCRIPTION
**Pull request by gefjon**
# Description of Changes

This PR implements a fast path for converting BFLATN -> BSATN for rows types which have a constant BSATN length.

This is done in terms of a `KnownBsatnLayout`, which stores the BSATN length of a type and a series of `memcpy` invocations to convert BFLATN into BSATN.

# API and ABI breaking changes

N/a

If this is an API or ABI breaking change, please apply the
corresponding GitHub label.

# Expected complexity level and risk

3 - some `unsafe`, some assumptions of BSATN and BFLATN formats.

# Testing

*Describe any testing you've done, and any testing you'd like your reviewers to do,
so that you're confident that all the changes work as expected!*

- [x] Wrote tests, incl. a proptest for consistency with the existing BFLATN -> BSATN conversion.
- [ ] Run benchmarks to determine that this is faster enough to be worthwhile.
  - This has to wait for me to actually integrate the new code into `table` and `core`.

### CODE REVIEW

File: crates/table/src/layout.rs [hunk 1]
```diff
@@ -22,12 +22,16 @@ use spacetimedb_sats::{
...
     } else {
-        let padding = required_alignment - misalignment;
-        base + padding
+        let misalignment = base % required_alignment;
+        if misalignment == 0 {
+            base
+        } else {
+            let padding = required_alignment - misalignment;
+            base + padding
+        }
```
Comment (gefjon) [inline crates/table/src/layout.rs:34]: This was panicking in my test case for `AlgebraicType::never()`, so I added a branch to avoid division by zero.

File: crates/table/src/bflatn_to_bsatn_fast_path.rs [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+        AlgebraicTypeLayout, HasLayout, PrimitiveType, ProductTypeElementLayout, ProductTypeLayout, RowTypeLayout,
+        SumTypeLayout, SumTypeVariantLayout,
+    },
+    util::{range_move, slice_assume_init_ref},
+};
+
+/// A precomputed BSATN layout for a type whose encoded length is a known constant,
+/// enabling fast BFLATN -> BSATN conversion.
+#[derive(PartialEq, Eq, Debug)]
+pub struct KnownBsatnLayout {
```
Comment (joshua-spacetime) [inline crates/table/src/bflatn_to_bsatn_fast_path.rs:27]: I might call this `StaticBsatnLayout` instead.

File: crates/table/src/bflatn_to_bsatn_fast_path.rs [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+            }
+        }
+    }
+
+    proptest! {
+        // The test `known_bsatn_same_as_bflatn_from` generates a lot of rejects,
+        // as a vast majority of the space of `ProductType` does not have a fixed BSATN length.
+        // Writing a proptest generator which produces only types that have a fixed BSATN length
+        // seems hard, because we'd have to generate sums with known matching layouts,
+        // so we just bump the `max_global_rejects` up as high as it'll go and move on with our lives.
```
Comment (joshua-spacetime) [inline crates/table/src/bflatn_to_bsatn_fast_path.rs:395]: I would prefer this comment to explain why this test is needed over `known_types_expected_layout`. Because from my reading, it doesn't seem like it gives us any further _observable_ test coverage.

File: crates/table/src/bflatn_to_bsatn_fast_path.rs [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+//! with no variable-length members.
+//! BFLATN is designed with this in mind, storing fixed-length portions of rows inline,
+//! at the expense of an indirection to reach var-length columns like strings.
+//! A majority of these types will also have a fixed BSATN length,
+//! but note that BSATN stores sum values (enums) without padding,
+//! so row types which contain sums may not have a fixed BSATN length
+//! if the sum's variants have different "live" unpadded lengths.
+//!
+//! For row types with fixed BSATN lengths, we can reduce the BFLATN -> BSATN conversion
+//! to a series of `memcpy`s, skipping over padding sequences.
```
[the remark this answers is not in the document]
Comment (gefjon) [reply crates/table/src/bflatn_to_bsatn_fast_path.rs:13]: I added:
```rust
//! This is potentially much faster than the more general  [`crate::bflatn_from::serialize_row_from_page`],
//! which traverses a [`RowTypeLayout`] and dispatches on the type of each column.
```
If you'd like more, let me know.

File: crates/table/src/bflatn_to_bsatn_fast_path.rs [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+    util::{range_move, slice_assume_init_ref},
+};
+
+/// A precomputed BSATN layout for a type whose encoded length is a known constant,
+/// enabling fast BFLATN -> BSATN conversion.
+#[derive(PartialEq, Eq, Debug)]
+pub struct KnownBsatnLayout {
+    /// The length of the encoded BSATN representation of a row of this type,
+    /// in bytes.
+    bsatn_length: u16,
```
[the remark this answers is not in the document]
Comment (gefjon) [reply crates/table/src/bflatn_to_bsatn_fast_path.rs]: Expanded comment with
```rust
/// Storing this allows us to pre-allocate correctly-sized buffers,
/// avoiding potentially-expensive `realloc`s.
```

File: crates/table/src/bflatn_to_bsatn_fast_path.rs [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+        self.fields.last().unwrap()
+    }
+
+    fn current_field_mut(&mut self) -> &mut MemcpyField {
+        self.fields.last_mut().unwrap()
+    }
+
+    fn next_bflatn_offset(&self) -> u16 {
+        let last = self.current_field();
+        last.bflatn_offset + last.length
```
[the remark this answers is not in the document]
Comment (gefjon) [reply crates/table/src/bflatn_to_bsatn_fast_path.rs:145]: No, padding can only be introduced by products, so we only have to handle it in `visit_product`/`visit_product_element`. I expanded the comment in `visit_product_element`:
```rust
// Note that this is the only place we have to reason about alignment and padding
// because the enclosing `ProductTypeLayout` has already computed valid aligned offsets
// for the elements.
```

File: crates/table/src/bflatn_to_bsatn_fast_path.rs (hunk 3 above)
Comment (gefjon) [reply crates/table/src/bflatn_to_bsatn_fast_path.rs:406]: Comment expanded; now in full reads:
```rust
// The test `known_bsatn_same_as_bflatn_from` generates a lot of rejects,
// as a vast majority of the space of `ProductType` does not have a fixed BSATN length.
// Writing a proptest generator which produces only types that have a fixed BSATN length
// seems hard, because we'd have to generate sums with known matching layouts,
// so we just bump the `max_global_rejects` up as high as it'll go and move on with our lives.
//
// Note that I (pgoldman 2024-03-21) tried modifying `generate_typed_row`
// to not emit `String`, `Array` or `Map` types (the trivially var-len types),
// but did not see a meaningful decrease in the number of rejects.
// This is because a majority of the var-len BSATN types in the `generate_typed_row` space
// are due to sums with inconsistent payload layouts.
//
// We still include the test `known_bsatn_same_as_bsatn_from`
// because it tests row types not covered in `known_types_expected_layout`,
// especially larger types with unusual sequences of aligned fields.
```

**Comment by gefjon** [comment]
I added a set of new benchmarks in `special`:

```
Benchmarking special/serialize/u32_u64_str/bflatn_to_bsatn_slow_path/count=100: Collecting 100 samples in estimated 5.special/serialize/u32_u64_str/bflatn_to_bsatn_slow_path/count=100
                        time:   [2.0599 µs 2.0639 µs 2.0686 µs]
                        thrpt:  [48.341 Melem/s 48.453 Melem/s 48.547 Melem/s]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
Benchmarking special/serialize/u32_u64_str/bflatn_to_bsatn_fast_path/count=100: Collecting 100 samples in estimated 5.special/serialize/u32_u64_str/bflatn_to_bsatn_fast_path/count=100
                        time:   [3.7009 µs 3.7067 µs 3.7132 µs]
                        thrpt:  [26.931 Melem/s 26.978 Melem/s 27.020 Melem/s]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

Benchmarking special/serialize/u32_u64_u64/bflatn_to_bsatn_slow_path/count=100: Collecting 100 samples in estimated 5.special/serialize/u32_u64_u64/bflatn_to_bsatn_slow_path/count=100
                        time:   [1.5816 µs 1.5841 µs 1.5870 µs]
                        thrpt:  [63.013 Melem/s 63.128 Melem/s 63.226 Melem/s]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
Benchmarking special/serialize/u32_u64_u64/bflatn_to_bsatn_fast_path/count=100: Collecting 100 samples in estimated 5.special/serialize/u32_u64_u64/bflatn_to_bsatn_fast_path/count=100
                        time:   [962.63 ns 964.98 ns 967.65 ns]
                        thrpt:  [103.34 Melem/s 103.63 Melem/s 103.88 Melem/s]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severespecial/serialize/u32_u64_str/bflatn_to_bsatn_slow_path/count=100
                        time:   [2.0835 µs 2.0875 µs 2.0919 µs]
                        thrpt:  [47.804 Melem/s 47.905 Melem/s 47.996 Melem/s]
                 change:
                        time:   [+0.5187% +0.8431% +1.1652%] (p = 0.00 < 0.05)
                        thrpt:  [-1.1518% -0.8361% -0.5160%]
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
Benchmarking special/serialize/u32_u64_str/bflatn_to_bsatn_fast_path/count=100: Collecting 100 samples in estimated 5.special/serialize/u32_u64_str/bflatn_to_bsatn_fast_path/count=100
                        time:   [2.2770 µs 2.2811 µs 2.2856 µs]
                        thrpt:  [43.752 Melem/s 43.839 Melem/s 43.917 Melem/s]
                 change:
                        time:   [-38.765% -38.583% -38.393%] (p = 0.00 < 0.05)
                        thrpt:  [+62.319% +62.821% +63.305%]
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high severe


```

Note that the "fast path" is applicable only to `u32_u64_u64`. I was surprised by how much slower the "fast path" is for `u32_u64_str`, and it turns out to be caused by:
```rust
            // Pre-allocate the capacity needed to write `result`.
            let len = bsatn::to_len(self)?;
            buf.reserve(len);
```
in `RowRef::to_bsatn_extend`. I took these lines from `instance_env.rs`, which always reserves capacity before serializing, but it turns out that commenting them out yields:
```rust
special/serialize/u32_u64_str/bflatn_to_bsatn_slow_path/count=100
                        time:   [2.0835 µs 2.0875 µs 2.0919 µs]
                        thrpt:  [47.804 Melem/s 47.905 Melem/s 47.996 Melem/s]
                 change:
                        time:   [+0.5187% +0.8431% +1.1652%] (p = 0.00 < 0.05)
                        thrpt:  [-1.1518% -0.8361% -0.5160%]
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
Benchmarking special/serialize/u32_u64_str/bflatn_to_bsatn_fast_path/count=100: Collecting 100 samples in estimated 5.special/serialize/u32_u64_str/bflatn_to_bsatn_fast_path/count=100
                        time:   [2.2770 µs 2.2811 µs 2.2856 µs]
                        thrpt:  [43.752 Melem/s 43.839 Melem/s 43.917 Melem/s]
                 change:
                        time:   [-38.765% -38.583% -38.393%] (p = 0.00 < 0.05)
                        thrpt:  [+62.319% +62.821% +63.305%]
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high severe
```

With those lines removed, the "fast path" is barely slower than the "slow path," which I suspect is caused by the extra branches.

Anyway, when the "fast path" applies, it's meaningfully faster than the "slow path," so I feel pretty good about this.

**Comment by gefjon** [comment]
I added a benchmark for `u64_u64_u32`, a type which is isomorphic to the existing `u32_u64_u64`, but has less interior padding. 
```
Benchmarking special/serialize/u64_u64_u32/bflatn_to_bsatn_slow_path/count=100: Collecting 100 samples in estimated 5.special/serialize/u64_u64_u32/bflatn_to_bsatn_slow_path/count=100
                        time:   [1.6070 µs 1.6102 µs 1.6139 µs]
                        thrpt:  [61.962 Melem/s 62.102 Melem/s 62.229 Melem/s]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) low mild
  1 (1.00%) high mild
Benchmarking special/serialize/u64_u64_u32/bflatn_to_bsatn_fast_path/count=100: Collecting 100 samples in estimated 5.special/serialize/u64_u64_u32/bflatn_to_bsatn_fast_path/count=100
                        time:   [678.13 ns 679.05 ns 680.07 ns]
                        thrpt:  [147.04 Melem/s 147.26 Melem/s 147.46 Melem/s]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) low mild
  1 (1.00%) high mild
```
The two have nearly identical "slow path" times, but the version without interior padding is nearly a third faster in the "fast path."

**Comment by gefjon** [comment]
Subscription benchmarks:
```
full-scan               time:   [64.706 ms 64.897 ms 65.092 ms]
                        change: [-19.002% -18.661% -18.299%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

full-join               time:   [470.04 µs 470.96 µs 472.04 µs]
                        change: [-6.5487% -6.2409% -5.9371%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

incr-select             time:   [327.52 ns 327.93 ns 328.37 ns]
                        change: [+6.1609% +6.3641% +6.5657%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

incr-join               time:   [2.0061 µs 2.0086 µs 2.0114 µs]
                        change: [-0.4693% -0.2668% -0.0684%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

query-indexes-multi     time:   [675.65 ns 677.30 ns 679.01 ns]
                        change: [-6.4458% -6.2716% -6.1022%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe
```
TL;DR:
- Big win for `full-scan`.
- Small win for `full-join`.
- Small loss for `incr-select`.
- No change for `incr-join`.
- Small win for `query-indexes-multi`.

I'll investigate the loss in `incr-select` and see if I can win  back any of that time.

**Comment by gefjon** [comment]
I'm stumped as to how this can be making `incr-select` slower, since the timing loop of that benchmark appears to never call `RowRef::to_bsatn`. Even weirder, I marked both new `RowRef` functions as `#[inline(never)]` (so that I could be sure they'd show up in a `perf` run; this is how I convinced myself that the benchmark didn't hit them) and the regression compared to master dropped to +1.8779, compared to +7.9029. I can only assume that I'm running up against the LLVM optimizer in some unfortunate way.

**Comment by joshua-spacetime** [comment]
@gefjon is the +6% for `incr-select` consistent? It's not uncommon for that one to swing by +/-5% on consecutive runs. It could be noise.

**Comment by gefjon** [comment]
> is the +6% for `incr-select` consistent? It's not uncommon for that one to swing by +/-5% on consecutive runs. It could be noise.

Yes, I ran two or three times and saw similar results.

**Comment by joshua-spacetime** [comment]
FWIW I get
```
full-scan               time:   [46.938 ms 47.206 ms 47.473 ms]
                        change: [-18.109% -17.639% -17.174%] (p = 0.00 < 0.05)
                        Performance has improved.

full-join               time:   [195.95 µs 196.07 µs 196.20 µs]
                        change: [-10.205% -10.038% -9.8600%] (p = 0.00 < 0.05)
                        Performance has improved.

incr-join               time:   [1.5647 µs 1.5665 µs 1.5686 µs]
                        change: [+1.0082% +1.2600% +1.4942%] (p = 0.00 < 0.05)
                        Performance has regressed.

incr-select             time:   [356.55 ns 356.69 ns 356.84 ns]
                        change: [+0.8042% +1.1048% +1.3752%] (p = 0.00 < 0.05)
                        Change within noise threshold.
```

File: crates/table/src/bflatn_to_bsatn_fast_path.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (joshua-spacetime) [reply crates/table/src/bflatn_to_bsatn_fast_path.rs:13]: Maybe a little more.

The reason this is faster, to my understanding, which may be incomplete, is that we can memcpy each bfatn field into a statically sized bsatn buffer. We can't memcpy in the general case because, in general, the size of the bsatn buffer cannot be computed statically.

Is this accurate?

> skipping over padding sequences

This is also a little confusing because I can't tell if you're implying that this contributes to the speed up because we can't skip over padding in the general case, or if this is just a constraint on the memcpy. I believe it's the latter, but wasn't entirely sure.

File: crates/table/src/bflatn_to_bsatn_fast_path.rs (hunk 3 above)
Comment (joshua-spacetime) [reply crates/table/src/bflatn_to_bsatn_fast_path.rs:406]: > // We still include the test `known_bsatn_same_as_bsatn_from`
// because it tests row types not covered in `known_types_expected_layout`,
// especially larger types with unusual sequences of aligned fields.

But are we guaranteed to catch any such failures?

File: crates/table/src/bflatn_to_bsatn_fast_path.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (gefjon) [reply crates/table/src/bflatn_to_bsatn_fast_path.rs:13]: What I'm trying to say here is that we can copy multiple fields in a single `memcpy` if they're stored without any intermediate padding in the BFLATN. So the fast path can do a single memcpy for `(u64, u64, u64)`, whereas the slow path will do 3 dispatches, 3 reads and 3 `serialize_u64`s.

File: crates/table/src/bflatn_to_bsatn_fast_path.rs (hunk 3 above)
Comment (gefjon) [reply crates/table/src/bflatn_to_bsatn_fast_path.rs:406]: This is really more of an existential question about how much you trust statistical testing. I believe this test will catch any such failures at a high rate.

**Comment by gefjon** [comment]
benchmarks please

File: crates/table/src/bflatn_to_bsatn_fast_path.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (gefjon) [reply crates/table/src/bflatn_to_bsatn_fast_path.rs:13]: Added:
```rust
//! For example, to serialize a row of type `(u64, u64, u32, u64)`,
//! [`bflatn_from`] will do four dispatches, three calls to `serialize_u64` and one to `serialize_u32`.
//! This module will make 2 `memcpy`s (or actually, `<[u8]>::copy_from_slice`s):
//! one of 20 bytes to copy the leading `(u64, u64, u32)`, which contains no padding,
//! and then one of 8 bytes to copy the trailing `u64`, skipping over 4 bytes of padding in between.
```

File: crates/table/src/bflatn_to_bsatn_fast_path.rs [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+    /// Length to `memcpy`, in bytes.
+    length: u16,
+}
+
+impl MemcpyField {
+    /// # Safety
+    ///
+    /// - `buf` must be at least `self.bsatn_offset + self.length` long.
+    /// - `row` must be at least `self.bflatn_offset + self.length` long.
+    /// - `row[self.bflatn_offset .. self.bflatn_offset + length]` must all be initialized.
```
[the remark this answers is not in the document]
Comment (gefjon) [reply crates/table/src/bflatn_to_bsatn_fast_path.rs:100]: I don't want readers of the comments to have to hunt down the definition of `range_move` in order to review safety.

File: crates/table/src/bflatn_to_bsatn_fast_path.rs [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+        let from = unsafe { slice_assume_init_ref(from) };
+        to.copy_from_slice(from);
+    }
+
+    fn is_empty(&self) -> bool {
+        self.length == 0
+    }
+}
+
+struct LayoutBuilder {
```
[the remark this answers is not in the document]
Comment (gefjon) [reply crates/table/src/bflatn_to_bsatn_fast_path.rs:116]: I think this comment, and other trivial comments (especially on non-`pub` items) are silly, but I'll include it anyways.

**Comment by bfops** [comment]
# Bot test results
**Conclusion**: :green_circle: Looks like a big improvement in `call_reducer`!

### Commits tested
before:
```
commit b5955edae4c27fa08f8ccb56df3dcaf1c7a57f6f
Merge: af73b29b 0ad01b77
Author: Zeke Foppa <196249+bfops@users.noreply.github.com>
Date:   Mon Mar 25 15:53:46 2024 -0700

    [bfops/perf-test-before]: Merge remote-tracking branch 'origin/bfops/enable-tracy' into bfops/perf-test-before

commit af73b29b5f687c95863772936dafaeea94c5ceaa
Author: Tyler Cloutier <cloutiertyler@aol.com>
Date:   Thu Mar 21 21:19:31 2024 -0700

    Made me the codeowner of the traits.rs file for the datastore
```
after:
```
commit d3b0008030f9739a012b1c3e407cb6b2e6e46140
Merge: ba8a8d93 0ad01b77
Author: Zeke Foppa <196249+bfops@users.noreply.github.com>
Date:   Mon Mar 25 15:55:53 2024 -0700

    [bfops/perf-test-after]: Merge remote-tracking branch 'origin/bfops/enable-tracy' into bfops/perf-test-after

commit ba8a8d93c321667615b185d79074cd6ed21e822a
Author: Phoebe Goldman <phoebe@clockworklabs.io>
Date:   Mon Mar 25 15:46:10 2024 -0400

    BFLATN -> BSATN fast-path for fixed-length rows (#1005)
    
    * Implement (but do not use) a fast path for BFLATN -> BSATN conversion
    
    * fmt and clippy
    
    * `u16` offset rather than `usize`
    
    * Address Joshua's review
    
    * Define methods on `RowRef` and `RelValue` which use the new serializer
    
    * Comment in `align_to` about div-by-zero
    
    Co-authored-by: Mazdak Farrokhzad <twingoow@gmail.com>
    Signed-off-by: Phoebe Goldman <phoebe@goldman-tribe.org>
    
    * Add benchmark comparing BFLATN -> BSATN with and without the fast path
    
    * Add benchmark on `u64_u64_u32`, which has less interior padding than `u32_u64_u64`
    
    * Remove `to_len` from `to_bsatn_extend`
    
    It turns out to be slower than just eating the `realloc`s.
    
    * Remove unused `to_bsatn_slice`
    
    I thought I would need it, but it ended up not being useful.
    
    * Expand comment with example; `Box<[...]>` to reduce memory footprint
    
    * Comments from Mazdak's review
    
    ---------
    
    Signed-off-by: Phoebe Goldman <phoebe@goldman-tribe.org>
    Co-authored-by: Mazdak Farrokhzad <twingoow@gmail.com>
```
### `call_reducer`
image

### `call_reducer_with_tx`
image

### `eval_incr`
image

### `eval_updates`
image

### `enemy_ai_agent_loop`
image

### `npc_ai_agent_loop`
image

**Comment by bfops** [comment]
Update: I believe I tested incorrect commits previously. I believe these results are correct:

### `call_reducer`
image

### `call_reducer_with_tx`
image

### `eval_incr`
image

### `eval_updates`
image

### `eval_binary`
image

### `enemy_ai_agent_loop`
image

### `npc_ai_agent_loop`
image

**Comment by bfops** [comment]
Correction: the first image (`call_reducer`) has a bug where the left and right traces are showing different source locations.

Here is a corrected screenshot:
image

There are still meaningful savings, but not the huge gap seen before (~10% faster).
